File: hearts/src/Game.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}

module Game where

import Polysemy
import Polysemy.Internal (send)
import qualified Polysemy.State as State
import Polysemy.State (State)
import Polysemy.Cont

import Data.IORef (IORef)
import qualified Data.IORef as IORef
import qualified Control.Monad as Monad
import GameEvent
import Cards
import Table
import Shuffle

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

import qualified Data.Set as Set
import Data.Set (Set)

import Debug.Trace (trace)

data GameStep monad a where
  BroadcastEvent :: GameEvent -> GameStep monad ()
  ReceiveCommand :: GameStep monad (Maybe GameCommand)

broadcastEvent :: Member GameStep effects => GameEvent -> Sem effects ()
broadcastEvent event = send (BroadcastEvent event)

receiveCommand :: Member GameStep effects => Sem effects (Maybe GameCommand)
receiveCommand = send ReceiveCommand

-- ein Event von den Spielern verarbeiten lassen
broadcastEvents ::
  (Member GameStep effects) =>
  [GameEvent] ->
  Sem effects ()
-- broadcastEvents events | trace ("broadcastEvents " ++ show events) False = undefined
broadcastEvents events =
  Monad.mapM_ broadcastEvent events

executeCommand ::
  (Member (State TableState) effects, Member GameStep effects) =>
  GameCommand ->
  Sem effects ([GameEvent], Maybe Player)
-- executeCommand command | trace ("executeCommand " ++ show command) False = undefined
executeCommand command =
  do state <- State.get
     let events = tableProcessCommand command state
     let state' = foldl (flip tableProcessEvent) state events
     State.put state'
     return (events, gameOver state')

playLoop ::
  (Member (State TableState) effects, Member GameStep effects) =>
  GameCommand ->
  Sem effects (Maybe Player)
-- playLoop command | trace ("playLoop " ++ (show command)) False = undefined
playLoop command =
  do
    (events, maybeWinner) <- executeCommand command
    broadcastEvents events
    case maybeWinner of
      Nothing ->
        do
          maybeCommand <- receiveCommand
          case maybeCommand of
            Just command -> playLoop command
            Nothing -> return Nothing
      Just winner -> 
        return (Just winner)
playGame ::
  (Member (State TableState) effects, Member GameStep effects) =>
  [Player] ->
  [Card] ->
  Sem effects (Maybe Player)
playGame players shuffledCards = do
  let hands = Map.fromList (zip players (map Set.fromList (distribute (length players) shuffledCards)))
  playLoop (DealHands hands)

data GameStepResult
  = EventBroadcast GameEvent (() -> Sem GameStepResultEffects GameStepResult)
  | WaitingForCommand (GameCommand -> Sem GameStepResultEffects GameStepResult)
  | GameDone (Maybe Player)

instance Show GameStepResult where
  show (EventBroadcast event _) = "EventBroadcast (" ++ (show event) ++ ")"
  show (WaitingForCommand _) = "WaitingForCommand"
  show (GameDone mbPlayer) = "GameDone (" ++ (show mbPlayer) ++ ")"

type GameStepResultEffects = '[GameStepResultEffect]
type GameStepResultEffect = Cont (Ref (Sem '[]) GameStepResult)

playOne :: (forall b. GameStepResult -> Sem GameStepResultEffects b) ->
   Sem (GameStep ': GameStepResultEffects) a -> Sem GameStepResultEffects a
playOne cont0 =
  interpret
    ( \step ->
        case step of
          BroadcastEvent event ->
            callCC
              ( \cont ->
                  cont0 (EventBroadcast event cont))
          ReceiveCommand ->
            callCC
              ( \cont ->
                  cont0 (WaitingForCommand (cont . Just))))

File: hearts-mike/src/main/scala/funar/hearts/Game.scala
package funar.hearts

import org.atnos.eff._
import org.atnos.eff.interpret._
import org.atnos.eff.all._
import org.atnos.eff.syntax.all._
import org.atnos.eff.addon.cats.effect._
import org.atnos.eff.syntax.addon.cats.effect._
import cats.Traverse
import cats.data._
import cats.implicits._
import funar.hearts.Table._
import cats.Eval

enum GameStep[+A] {
  case BroadcastEvent(event: GameEvent) extends GameStep[Unit]
  case ReceiveCommand extends GameStep[Option[GameCommand]]
}

enum GameStepResult[A] {
  case EventBroadcast(event: GameEvent, continuation: Unit => GameStepResult[A])
  case WaitingForCommand(continutation: GameCommand => GameStepResult[A])
  case GameDone(result: A)
}

object Game {
  import GameStep._

  type _gameStep[Effects] = GameStep |= Effects
  type _tableState[Effects] = State[TableState, _] |= Effects

  def broadcastEvent[T, R : _gameStep](event: GameEvent): Eff[R, Unit] =
    Eff.send[GameStep, R, Unit](BroadcastEvent(event))

  def receiveCommand[T, R : _gameStep]: Eff[R, Option[GameCommand]] =
    Eff.send[GameStep, R, Option[GameCommand]](ReceiveCommand)

  // ein Event von den Spielern verarbeiten lassen
  def broadcastEvents[Effects: _gameStep](events: Seq[GameEvent]): Eff[Effects, Unit] =
    events.traverse_(broadcastEvent)

  def executeCommand[Effects: _gameStep : _tableState](command: GameCommand): Eff[Effects, (Seq[GameEvent], Option[Player])] =
    for {
      state <- get[Effects, TableState]
      events = tableProcessCommand(command, state)
      state1 = events.foldLeft(state) { (state, event) => tableProcessEvent(event, state) } 
      _ <- put[Effects, TableState](state1)
    } yield (events, gameOver(state1))

  def playLoop[Effects: _gameStep : _tableState](command: GameCommand): Eff[Effects, Option[Player]] =
    for {
      pair <- executeCommand(command)
      (events, maybeWinner) = pair
      _ <- broadcastEvents(events)
      res <- maybeWinner match {
        case None =>
          for {
            maybeCommand <- receiveCommand
            _ = println("received command " ++ maybeCommand.toString)
            res <- maybeCommand match {
              case Some(command) => playLoop(command)
              case None => Eff.pure(None)
            }
          } yield res
        case Some(winner) =>
          Eff.pure(Some(winner))
      }
    } yield res

  def playGame[Effects: _gameStep : _tableState](players: Seq[Player], deck: List[Card]): Eff[Effects, Option[Player]] = {
    val hands = players.zip(distribute(players.length, deck).map(_.toSet)).toMap
    playLoop(GameCommand.DealHands(hands))
  }

  def playOne[A](effects: Eff[Fx1[GameStep], A]): GameStepResult[A] = {
    val interpreter = new Interpreter[GameStep, NoFx, A, GameStepResult[A]] {
      import GameStepResult._
      override def onPure(a: A) = Eff.pure{GameDone(a)}
      override def onEffect[X](x: GameStep[X], continuation: Continuation[NoFx, X, GameStepResult[A]]) =
        x match {
          case BroadcastEvent(event) =>
            Eff.pure(EventBroadcast(event, { _ => continuation(()).run }))
          case ReceiveCommand =>
            Eff.pure(WaitingForCommand({ command => continuation(Some(command)).run }))
        }
      override def onLastEffect[X](x: GameStep[X], continuation: Continuation[NoFx, X, Unit]): Eff[NoFx, Unit] =
        Eff.pure(()) // ??? rilly?
      override def onApplicativeEffect[X, T[_] : Traverse](xs: T[GameStep[X]], continuation: Continuation[NoFx, T[X], GameStepResult[A]]): Eff[NoFx, GameStepResult[A]] = {
        xs.toList match {
          case Nil => continuation(Vector.empty[X].asInstanceOf[T[X]]) // we know it's a vector
          case BroadcastEvent(event)::rest =>
            Eff.pure(EventBroadcast(event,
                      {_ => onApplicativeEffect(rest.toVector, Continuation.lift { (rest: Vector[X]) => continuation((() +: rest).toVector.asInstanceOf[T[X]]) }).run }))
          case ReceiveCommand::rest =>
            Eff.pure(WaitingForCommand({ command => 
              onApplicativeEffect(rest.toVector, Continuation.lift { (rest: Vector[X]) => continuation((Some(command) +: rest).toVector.asInstanceOf[T[X]]) }).run }))
        }
      }
    }
    runInterpreter(effects)(interpreter).run
  }


  def distribute[A](n: Int, list: List[A]): Seq[List[A]] = {
    def extract[A](list: List[A]): List[A] =
      list match {
        case Nil => Nil
        case (x :: xs) => 
          x :: extract(xs.drop(n-1))
      }
    (0 until n).map { i =>
      extract(list.drop(i))
    }
  }
    
}